### DISCUSSION TITLE
Build for iOS works, running on iOS Simulator get "Failed to load vulkan library!"

### DISCUSSION THREAD

**Discussion by klmodrow** [KhronosGroup/Vulkan-Samples#1135] [Q&A]
I am trying to get the samples to run on iOS Simulator on my MacBook through Xcode 15.4. I can get everything to work fine by following the instructions on this repo and a few other resources, up until I go to run the application with different samples. I can build and run the application but almost none of the samples work. The only sample I can seem to run without errors is hello_triangle in samples/api. The sample hpp_hello_triangle for example runs and prints the following to the terminal:
```
 [error] Error Message: Failed to load vulkan library!
```
And after some investigation, it seems that the hello_triangle sample (the one that works) goes to initialize volk, and loads the vulkan library by searching for different library names, supporting the Frameworks path within the application bundle:
```
VkResult volkInitialize(void){

[...]

#elif defined(__APPLE__)
	void* module = dlopen("libvulkan.dylib", RTLD_NOW | RTLD_LOCAL);
	if (!module)
		module = dlopen("libvulkan.1.dylib", RTLD_NOW | RTLD_LOCAL);
	if (!module)
		module = dlopen("libMoltenVK.dylib", RTLD_NOW | RTLD_LOCAL);
    // Add support for using Vulkan and MoltenVK in a Framework. App store rules for iOS
    // strictly enforce no .dylib's. If they aren't found it just falls through
    if (!module)
        module = dlopen("vulkan.framework/vulkan", RTLD_NOW | RTLD_LOCAL);
    if (!module)
        module = dlopen("MoltenVK.framework/MoltenVK", RTLD_NOW | RTLD_LOCAL);
	// modern versions of macOS don't search /usr/local/lib automatically contrary to what man dlopen says
	// Vulkan SDK uses this as the system-wide installation location, so we're going to fallback to this if all else fails
	if (!module && getenv("DYLD_FALLBACK_LIBRARY_PATH") == NULL)
		module = dlopen("/usr/local/lib/libvulkan.dylib", RTLD_NOW | RTLD_LOCAL);
	if (!module)
		return VK_ERROR_INITIALIZATION_FAILED;

	vkGetInstanceProcAddr = (PFN_vkGetInstanceProcAddr)dlsym(module, "vkGetInstanceProcAddr");
#else

[...]

	loadedModule = module;
	volkGenLoadLoader(NULL, vkGetInstanceProcAddrStub);

	return VK_SUCCESS;
}
```
And the hpp_hello_triangle sample (the one that doesn't work) uses the vulkan.hpp header file's DynamicLoader class to load the vulkan library when creating an instance, but cannot find it because it only tries finding the library with a .dylib extension and has no sense of the Frameworks path included in an iOS app bundle (see defined(\_\_APPLE\_\_) sections):
```
class DynamicLoader
{
public:
#  ifdef VULKAN_HPP_NO_EXCEPTIONS
    DynamicLoader( std::string const & vulkanLibraryName = {} ) VULKAN_HPP_NOEXCEPT
#  else
    DynamicLoader( std::string const & vulkanLibraryName = {} )
#  endif
    {
      if ( !vulkanLibraryName.empty() )
      {
#  if defined( __unix__ ) || defined( __APPLE__ ) || defined( __QNX__ ) || defined( __Fuchsia__ )
        m_library = dlopen( vulkanLibraryName.c_str(), RTLD_NOW | RTLD_LOCAL );
#  elif defined( _WIN32 )
        m_library = ::LoadLibraryA( vulkanLibraryName.c_str() );
#  else
#    error unsupported platform
#  endif
      }
      else
      {
#  if defined( __unix__ ) || defined( __QNX__ ) || defined( __Fuchsia__ )

        [...]

#  elif defined( __APPLE__ )
        m_library = dlopen( "libvulkan.dylib", RTLD_NOW | RTLD_LOCAL );
        if ( m_library == nullptr )
        {
          m_library = dlopen( "libvulkan.1.dylib", RTLD_NOW | RTLD_LOCAL );
        }
#  elif defined( _WIN32 )
        
        [...]

#  else
#    error unsupported platform
#  endif
      }

#  ifndef VULKAN_HPP_NO_EXCEPTIONS
      if ( m_library == nullptr )
      {
        // NOTE there should be an InitializationFailedError, but msvc insists on the symbol does not exist within the scope of this function.
        throw std::runtime_error( "Failed to load vulkan library!" );
      }
#  endif
}
```
And because m_library will be NULL since the library is in the Frameworks folder in the application bundle, it throws the runtime error. Now, I know iOS doesn't allow the use of .dylib libraries, so that leads me to think I have the project in the wrong configuration or that all the samples that error out like this just aren't supported yet. 

Can anyone help me understand how to get the iOS version of the samples to work or if they are even suppose to? How can I edit the Xcode project build for iOS to not assume system libraries? Is this expected behavior since using the Vulkan headers assumes using .dylib libraries? Any insight would help me greatly as this is my first time working with iOS applications and Vulkan. Thanks in advance!

**Comment by gpx1000** [comment]
There's a bug in vulkan.hpp and how it loads the library from the loader that doesn't exist in volk.  When I originally worked on getting iOS support, I remember running into this.  I coded around it by explicitly telling vulkan.hpp to use the framework.  https://github.com/KhronosGroup/Vulkan-Samples/blob/fdce530c029514e2f66296288f9ee26a05058ec1/framework/vulkan_sample.h#L1012  The *real* fix would be to submit an issue to vulkan.hpp and do a PR there that does the same thing that volk does.  I dropped the ball on remembering to get that fix submitted to vulkan.hpp

The local fix here could be to add TARGET_OS_SIMULATOR to the check on system type.

**Reply by klmodrow** [reply]
Thanks for the swift reply and the information. Would you be able to elaborate more on the local fix recommendation you mentioned? Where could I add TARGET_OS_SIMULATOR?

**Comment by klmodrow** [comment]
The partial "fix" for running the samples out of the box for iOS-Simulator ended up being a change to the explicit file name that @gpx1000 mentioned [above ](https://github.com/KhronosGroup/Vulkan-Samples/discussions/1135#discussioncomment-10413429)within Vulkan-Samples/framework/vulkan_sample.h from "vulkan.framework/vulkan" to "MoltenVK.framework/MoltenVK" under the #if TARGET_OS_IPHONE check. This is a quick fix and still does not allow a lot of the samples to run, just more than only hello_triangle when you build and run for iOS-Sim out of the box.

**Reply by SRSaunders** [reply]
I think **vulkan.hpp** is actually fine as is.  There are two issues here that will solve it:

1. **global_options.cmake** line 39 currently is:

`if(USE_MoltenVK OR (IOS AND (NOT Vulkan_MoltenVK_FOUND OR (${CMAKE_OSX_SYSROOT} STREQUAL "iphonesimulator" AND ${CMAKE_HOST_SYSTEM_PROCESSOR} STREQUAL "x86_64"))))`

Instead it should be:

`if(USE_MoltenVK OR (IOS AND (NOT Vulkan_MoltenVK_FOUND OR ${CMAKE_OSX_SYSROOT} STREQUAL "iphonesimulator")))`

This will force MoltenVK.xcframework to be used which contains a build for ios-arm64_x86_64-simulator.  The regular vulkan framework does not.  When I first wrote this code I was testing mainly on x86_64 hosts and missed the arm64 case. Sorry.

2. If you are running on arm64 (Apple Silicon), you will likely have to sign the binary.  The docs should be updated to read:

"Alternatively, you can build for the iOS Simulator without code signing or specifying a bundle identifier (a default bundle id will be used). However, depending on your host architecture, you MUST select either arm64 (Apple Silicon) or x86_64 in the command below. _Note: On arm64 hosts the Vulkan library may not load on iOS Simulator without specifying your TeamID and bundle identifier as shown above._"

I have a PR planned for both of these things, but was waiting for @gpx1000 to respond on issue #1397 so I could solve both things at once.  @klmodrow please try steps 1 and 2 above and let me know.  Many of the samples do run on the Simulator, but some do not due to lack of feature support (e.g. descriptor indexing).  Support is broader on physical iOS devices.

***I just realized this is a very old thread - @klmodrow please ignore if no longer relevant for you.***
